pyomo/contrib/gdpbb/GDPbb.py

```diff
@@ -0,0 +1,252 @@
+        always be available, and so this should reflect that possibility.
+
+        """
+        return True
+
+    def solve(self, model, **kwds):
+        config = self.CONFIG(kwds.pop('options', {}))
+        config.set_value(kwds)
+
+        # Validate model to be used with gdpbb
+        self.validate_model(model)
+        # Set solver as an MINLP
+        solver = SolverFactory(config.solver)
+        solve_data = GDPbbSolveData()
+        solve_data.timing = Container()
+        solve_data.original_model = model
+        solve_data.results = SolverResults()
+
+        old_logger_level = config.logger.getEffectiveLevel()
+        with time_code(solve_data.timing, 'total'),\
+                restore_logger_level(config.logger),\
+                create_utility_block(model, 'GDPbb_utils', solve_data):
+            if config.tee and old_logger_level > logging.INFO:
+                # If the logger does not already include INFO, include it.
+                config.logger.setLevel(logging.INFO)
+
+            #Setup results
+            self.setup_results_object(solve_data,model,config)
+            # Initialize list containing indicator vars for reupdating model after solving
+            indicator_list_name = unique_component_name(model, "_indicator_list")
+            indicator_vars = []
+            for disjunction in model.component_data_objects(
+                    ctype=Disjunction, active=True):
+                for disjunct in disjunction.disjuncts:
+                    indicator_vars.append(disjunct.indicator_var)
+            setattr(model, indicator_list_name, indicator_vars)
+
+            # get objective sense
+            objectives = model.component_data_objects(Objective, active=True)
+            obj = next(objectives, None)
+            obj_sign = 1 if obj.sense == minimize else -1
+            solve_data.results.problem.sense = obj.sense
+            # clone original model for root node of branch and bound
+            root = model.clone()
+
+            # set up lists to keep track of which disjunctions have been covered.
+
+            # this list keeps track of the original disjunctions that were active and are soon to be inactive
+            root.GDPbb_utils.unenforced_disjunctions = list(
+                disjunction for disjunction in root.GDPbb_utils.disjunction_list if disjunction.active
+            )
+
+            # this list keeps track of the disjunctions that have been activated by the branch and bound
+            root.GDPbb_utils.curr_active_disjunctions = []
+
+            # deactivate all disjunctions in the model
+            # self.indicate(root)
+            for djn in root.GDPbb_utils.unenforced_disjunctions:
+                djn.deactivate()
+            # Deactivate all disjuncts in model. To be reactivated when disjunction
+            # is reactivated.
+            for disj in root.component_data_objects(Disjunct, active=True):
+                disj._deactivate_without_fixing_indicator()
```

Comment:
Design question for @sjkale, @qtothec, and @emma58: now that we have algorithms relying on this method, should we move to make it a public, officially supported method?

Reply:
I think so. I was never not entirely comfortable with the current `deactivate()` fixes indicator variable to zero scheme, since deactivation feels like it ought to be a problem relaxation, but it can often end up being a problem restriction, as deactivating all the disjuncts in a disjunction yields an infeasible problem.

Maybe we can call it `.relax()`?

Reply:
I could be OK with `.relax()` (although `relax_disjunct()` might be better).  I think it makes a lot of sense for `.deactivate()` to fix the indicator to 0: if you are removing a disjunct from a model, then the solver should NOT be allowed to select it's indicator variable as a valid solution to the disjunction.  Can you remind me the situation where you want the disjunct removed, but the indicator variable still in the model?  Maybe there is a different API that we should be considering (a `relax` method on a Disjunction?)?

Reply:
Because I had originally seen `deactivate()` as implying relaxation. By that line of argument, deactivating a `Block` should somehow remove all of its variables from the model? (this thought experiment again)

Reply:
That's a tough question.  It depends on if a model is defined by it's active constraints, or by everything reachable through an active block path.  The problem is that there is correlated information: the "activeness" of a variable and the use of a variable in an active constraint.  For general variables, one can make an argument that deactivating the block shouldn't effect (restrict) it's use in other constraints.  Now, @carldlaird has pointed out that this problem goes away if variables are not allowed to be referenced outside of the block in which it is attached (and any sub-blocks).  In that case, deactivating the block with the variable will (by definition) deactivate all related constraints.  The downside is that it becomes challenging both to enforce, and to create models (users would have a lot a=b constraints, plus all the linking constraints would be at the sub-block level, which violates encapsulation ... making building up larger models from components more challenging).

For disjuncts it's even bit trickier: the variable *indicates* if the associated constraints are enforced.  Deactivating the block with the constraints, but allowing the indicator variable to be selected violate that semantic "promise".  But, as the indicator variable is typically used in a Disjunction, that relationship is always expressed *outside* the scope of any of the disjuncts.  As the constraint is formed as a transformation, it is easy to exclude the indicator variables from deactivated disjuncts (effectively implying that they are fixed to 0 in that constraint).  However, that does not help other logical constraints that may reference the deactivated disjunct (by leaving the indicator var free, you could see a solver setting it to 1 to satisfy a logical side-constraint (from `Omega(Y)==True`), but not have selected that disjunct in the disjunction.